### PR TITLE
[ValueTracking] Handle intrinsics in `computeConstantRange`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch is needed by https://github.com/llvm/llvm-project/issues/95255.

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=164a57805bd06f8857730b104fe4e4f11ed83ba1&to=bec349692a0c7f667a3370d14d8fbae0d5e2ab31&stat=instructions:u

**Comment by dtcxzyw** [comment]
> Failed Tests (1):
  LLVM :: CodeGen/AMDGPU/select-constant-cttz.ll

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -9687,12 +9687,12 @@ ConstantRange llvm::computeConstantRange(const Value *V, bool ForSigned,
                                          unsigned Depth) {
   assert(V->getType()->isIntOrIntVectorTy() && "Expected integer instruction");
 
-  if (Depth == MaxAnalysisRecursionDepth)
-    return ConstantRange::getFull(V->getType()->getScalarSizeInBits());
-
   if (auto *C = dyn_cast<Constant>(V))
     return C->toConstantRange();
 
+  if (Depth == MaxAnalysisRecursionDepth)
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:9693]: Related test: `@immarg_test`

File: llvm/test/Analysis/ValueTracking/constant-ranges.ll [hunk 2]
```diff
@@ -230,3 +230,72 @@ define i1 @srem_negC_fail1(i8 %x) {
...
+;
+  %sh_prom = zext nneg i32 %y to i64
+  %shl = shl nuw i64 1, %sh_prom
+  %cmp1 = icmp eq i64 %z, 0
+  %umin1 = call i64 @llvm.umin.i64(i64 %shl, i64 %z)
+  %sel = select i1 %cmp1, i64 1, i64 %umin1
+  %umin2 = call i64 @llvm.umin.i64(i64 %x, i64 %sel)
+  %cmp = icmp ugt i64 %umin2, -71777214294589697
+  ret i1 %cmp
+}
```
Comment (nikic) [inline llvm/test/Analysis/ValueTracking/constant-ranges.ll:246]: Won't this case already be handled by CVP with the shl nuw support from https://github.com/llvm/llvm-project/pull/100594?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -9728,6 +9728,17 @@ ConstantRange llvm::computeConstantRange(const Value *V, bool ForSigned,
...
+    if (const auto *II = dyn_cast<IntrinsicInst>(V)) {
+      Intrinsic::ID IID = II->getIntrinsicID();
+      if (ConstantRange::isIntrinsicSupported(IID)) {
+        SmallVector<ConstantRange, 2> OpRanges;
+        for (Value *Op : II->args())
+          OpRanges.push_back(computeConstantRange(Op, ForSigned, UseInstrInfo,
+                                                  AC, CtxI, DT, Depth + 1));
+        CR = CR.intersectWith(ConstantRange::intrinsic(IID, OpRanges));
+      }
+    }
```
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:9741]: Placing the code here makes no sense, there is already getRangeForIntrinsic above.

**Review by nikic — COMMENTED** [review]
> This patch is needed by https://github.com/llvm/llvm-project/issues/95255.

How/why does that patch need this? I don't see the relation at all.

**Review by nikic — CHANGES_REQUESTED** [review]
Independent of the specific feedback, I'm rejecting this in principle. See this comment in the doc comment:

> This is intended as a cheap, non-recursive check.

Complex recursive cases are intended to be handled by SCCP or LVI/CVP.

I *am* willing to explore making computeConstantRange() a full-blown recursive analysis, but only if this is done in a principled fashion, and not by adding these things piece-meal, such that each step looks free but we have no insight into the overall cost.

That is, if we want to go in this direction, computeConstantRange() needs to be switched fully to a recursive implementation, so that we can properly evaluate the cost/benefit of doing so. Maybe it's fine. Maybe it's fine if we switch some of the existing calls to use maximum starting depth. Maybe we can get rid of computeConstantRangeIncludingKnownBits() and use *only* computeConstantRange() (possibly having cross-recursion between the two to pick whichever is best for a given opcode).

**Comment by dtcxzyw** [comment]
My original plan is to set range metadata for `ctpop` in InstCombine: https://github.com/llvm/llvm-project/issues/95255#issuecomment-2166944909
https://github.com/llvm/llvm-project/blob/154d00d5d0416c42388d6e82cac96fbc091101d2/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp#L694-L701
```
define i1 @src(i32 noundef %n) {
entry:
  %tobool = icmp ne i32 %n, 0
  call void @llvm.assume(i1 %tobool)
  %0 = tail call i32 @llvm.ctpop.i32(i32 %n)
  %cmp = icmp eq i32 %0, 1
  ret i1 %cmp
}

define i1 @tgt(i32 noundef %n) {
entry:
  %tobool = icmp ne i32 %n, 0
  call void @llvm.assume(i1 %tobool)
  %0 = tail call range(i32 1, 33) i32 @llvm.ctpop.i32(i32 %n)
  %cmp = icmp eq i32 %0, 1
  ret i1 %cmp
}
```
Is it ok to infer range metadata for `ctpop` in CVP/SCCP?

**Comment by tschuett** [comment]
Or recursive and caching:
https://developers.redhat.com/blog/2019/10/11/a-upside-down-approach-to-gcc-optimizations

**Comment by andjo403** [comment]
for the https://github.com/llvm/llvm-project/issues/95255#issuecomment-2166944909 is it not enough to do a non Zero check something like this in the foldCtpop
```
    unsigned KnownMinPop = Known.countMinPopulation();
    unsigned Lower =
        KnownMinPop ? KnownMinPop
                    : isKnownNonZero(
                          Op0, IC.getSimplifyQuery().getWithInstruction(&II));
    ConstantRange Range(APInt(BitWidth, Lower),
                        APInt(BitWidth, Known.countMaxPopulation() + 1));
```

**Comment by dtcxzyw** [comment]
> for the [#95255 (comment)](https://github.com/llvm/llvm-project/issues/95255#issuecomment-2166944909) is it not enough to do a non Zero check something like this in the foldCtpop
> 
> ```
>     unsigned KnownMinPop = Known.countMinPopulation();
>     unsigned Lower =
>         KnownMinPop ? KnownMinPop
>                     : isKnownNonZero(
>                           Op0, IC.getSimplifyQuery().getWithInstruction(&II));
>     ConstantRange Range(APInt(BitWidth, Lower),
>                         APInt(BitWidth, Known.countMaxPopulation() + 1));
> ```

Are you willing to file a PR with this method? I drafted an alternative patch before submitting this PR:
```
diff --git a/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp b/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
index f6c4b6e18093..9f04ed85e831 100644
--- a/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
+++ b/llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp
@@ -692,12 +692,17 @@ static Instruction *foldCtpop(IntrinsicInst &II, InstCombinerImpl &IC) {
                             Ty);
 
   // Add range attribute since known bits can't completely reflect what we know.
-  if (BitWidth != 1 && !II.hasRetAttr(Attribute::Range) &&
-      !II.getMetadata(LLVMContext::MD_range)) {
-    ConstantRange Range(APInt(BitWidth, Known.countMinPopulation()),
-                        APInt(BitWidth, Known.countMaxPopulation() + 1));
-    II.addRangeRetAttr(Range);
-    return &II;
+  if (BitWidth != 1) {
+    ConstantRange OldRange = II.getRange().value_or(ConstantRange::getNonEmpty(APInt::getZero(BitWidth),
+                                      APInt(BitWidth, BitWidth + 1)));
+    ConstantRange Range = OldRange;
+    Range = Range.intersectWith(ConstantRange(APInt(BitWidth, Known.countMinPopulation()), APInt(BitWidth, Known.countMaxPopulation() + 1)), ConstantRange::Unsigned);
+    if (Range.contains(APInt::getZero(BitWidth)) && isKnownNonZero(Op0, IC.getSimplifyQuery().getWithInstruction(&II)))
+      Range = Range.intersectWith(ConstantRange(APInt(BitWidth, 1), APInt(BitWidth, BitWidth + 1)), ConstantRange::Unsigned);
+    if (Range != OldRange) {
+      II.addRangeRetAttr(Range);
+      return &II;
+    }
   }
 
   return nullptr;
```

**Comment by andjo403** [comment]
sure can make a PR

**Pull request closed without merging**